src/main/java/edu/cornell/kfs/pdp/batch/service/impl/ProcessPdpCancelPaidHelperServiceImpl.java

```diff
@@ -0,0 +1,149 @@
+
+    /**
+     * Default implementation uses most of the "while" loop contents from the ProcessPdpCancelPaidServiceImpl.processPdpCancels method,
+     * with additional CU-related changes as needed.
+     */
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
+    @Override
+    public void processPdpCancel(PaymentDetail paymentDetail, Date processDate) {
+        String documentTypeCode = paymentDetail.getFinancialDocumentTypeCode();
+        String documentNumber = paymentDetail.getCustPaymentDocNbr();
+
+        boolean primaryCancel = paymentDetail.getPrimaryCancelledPayment();
+        boolean disbursedPayment = PdpConstants.PaymentStatusCodes.CANCEL_PAYMENT.equals(paymentDetail.getPaymentGroup().getPaymentStatusCode());
+        
+        //KFSPTS-2719
+        boolean crCancel = false;
+        PaymentDetailExtendedAttribute paymentDetailExtendedAttribute = (PaymentDetailExtendedAttribute) paymentDetail.getExtension();
+        if (ObjectUtils.isNotNull(paymentDetailExtendedAttribute)) {
+            crCancel = paymentDetailExtendedAttribute.getCrCancelledPayment();
+        }
+
+        if (purchasingAccountsPayableModuleService.isPurchasingBatchDocument(documentTypeCode)) {
+            ((CuPurchasingAccountsPayableModuleService) purchasingAccountsPayableModuleService).handlePurchasingBatchCancels(
+                    documentNumber, documentTypeCode, primaryCancel, disbursedPayment, crCancel);
+        }
+        else {
+            PaymentSourceToExtractService<PaymentSource> extractService = getPaymentSourceToExtractService(paymentDetail);
+            if (extractService != null) {
+                try {
+                    PaymentSource dv = (PaymentSource) documentService.getByDocumentHeaderId(documentNumber);
+                    if (dv != null) {
+                        if (disbursedPayment || primaryCancel || crCancel) {
+                            if (!crCancel) {
+                                extractService.cancelPayment(dv, processDate);
+                            }
+                        } else {
+                            extractService.resetFromExtraction(dv);
+                        }
+                    }
+                } catch (WorkflowException we) {
+                    throw new RuntimeException("Could not retrieve document #" + documentNumber, we);
+                }
+            } else {
+                LOG.warn("processPdpCancel() Unknown document type (" + documentTypeCode + ") for document ID: " + documentNumber);
+                return;
+            }
+        }
+
+        paymentGroupService.processCancelledGroup(paymentDetail.getPaymentGroup(), processDate);
+    }
+
+    /**
+     * Default implementation uses most of the "while" loop contents from the ProcessPdpCancelPaidServiceImpl.processPdpPaids method,
+     * with additional CU-related changes as needed.
+     */
+    @Transactional(propagation = Propagation.REQUIRES_NEW)

```

Comment:
I have looked a bit into Propagation.REQUIRES_NEW and it looks like it might cause some performance overhead. This job already has some performance issues when we have large amounts of data to process. We should also do some testing to see how performance is affected.

Reply:
In that case, I could try changing this so that the REQUIRES_NEW processing will handle batches of a specific size, rather than processing each document individually. That should help improve efficiency without rolling back the entire set of updates on each job failure.

Reply:
After further discussion with Daniela, I will leave the per-payment transaction setup in place until we can get a better idea of how much of a performance impact it causes. (This may have to wait until the code moves to DEV, given the inherent slowness issues when running KFS locally.) If it ends up causing too much of a slow-down, then I can come up with a batch-processing-based solution.